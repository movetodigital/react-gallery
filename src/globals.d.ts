declare module '@storybook/addon-console';
