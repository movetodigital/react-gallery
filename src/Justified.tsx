import * as React from 'react';
import styled from 'styled-components';

type FunctionName = (n: object) => any;

interface ImageItemProps {
  src?: string;
  title?: string;
  alt?: string;
}

interface ComponentProps {
  images: Array<ImageItemProps>;
  gridColumns: number;
  gridGutter: number;
  aspectRatio?: number;
  imageRenderer?: FunctionName;
}

interface StyleItemProps {
  readonly width: number;
  readonly padding: string;
  readonly ratio: number;
}

interface RendererProps {
  props: ImageItemProps;
  key: number;
}

const Images = styled.div`
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
`;

const Item = styled.div<StyleItemProps>`
  color: white;
  position: relative;
  box-sizing: border-box;
  width: ${props => props.width}%;
  margin: ${props => props.padding};

  > div {
    width: 100%;
    padding-top: ${props => props.ratio}%;
  }
`;

const DefaultImage = styled.img`
  width: 100%;
  height: 100%;
  object-position: 50% 50%;
  object-fit: cover;
`;

const DefaultImageContainer = styled.div`
  position: absolute;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
`;

const defaultRender = ({props: {src, title, alt}, key}: RendererProps) => {
  return (
    <DefaultImageContainer>
      <DefaultImage key={src} src={src} title={title} alt={alt} />
    </DefaultImageContainer>
  );
};

const Justified: React.FunctionComponent<ComponentProps> = ({
  images,
  gridColumns,
  gridGutter,
  aspectRatio = 1,
  imageRenderer = defaultRender,
}) => {
  const gutter = gridGutter / 2;
  const width = (100 - (gridColumns - 1) * gutter * 2) / gridColumns;
  const ratio = 100 / aspectRatio;

  return (
    <Images>
      {images.map((el, index) => (
        <Item
          key={`${index}_${gutter}`}
          width={width}
          padding={`${gutter}% ${
            (index + 1) % gridColumns === 0 ? '0' : gutter
          }% ${gutter}% ${index % gridColumns === 0 ? '0' : gutter}%`}
          ratio={ratio}
        >
          <div>{imageRenderer({props: el, key: index})}</div>
        </Item>
      ))}
    </Images>
  );
};

export {Justified};
