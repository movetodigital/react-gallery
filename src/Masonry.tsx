import * as React from 'react';
import styled from 'styled-components';

type FunctionName = (n: object) => any;

interface ImageItemProps {
  src?: string;
  title?: string;
  alt?: string;
}

interface ComponentProps {
  images: Array<ImageItemProps>;
  gridColumns: number;
  gridGutter: number;
  imageRenderer?: FunctionName;
}

interface ImagesComponentProps {
  gridGutter: number;
  gridColumns: number;
}

interface StyleItemProps {
  gridGutter: number;
}

const Images = styled.div<ImagesComponentProps>`
  column-count: ${props => props.gridColumns};
  margin: 0 auto;
  column-gap: ${props => props.gridGutter}%;
`;

const Item = styled.div<StyleItemProps>`
  margin: 0 0 ${props => props.gridGutter}em;
`;

const DefaultImage = styled.img`
  max-width: 100%;
  max-height: 100%;
`;

interface RendererProps {
  props: ImageItemProps;
  key: number;
}

const defaultRender = ({props: {src, title, alt}, key}: RendererProps) => {
  return <DefaultImage key={src} src={src} title={title} alt={alt} />;
};

const Masonry: React.FunctionComponent<ComponentProps> = ({
  images,
  gridColumns,
  gridGutter,
  imageRenderer = defaultRender,
}) => {
  return (
    <Images gridColumns={gridColumns} gridGutter={gridGutter}>
      {images.map((el, index) => (
        <Item key={`${index}_${gridGutter}`} gridGutter={gridGutter}>
          <div>{imageRenderer({props: el, key: index})}</div>
        </Item>
      ))}
    </Images>
  );
};

export {Masonry};
