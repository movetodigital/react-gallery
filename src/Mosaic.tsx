import * as React from 'react';

interface Props {
  images: Array<string>;
}

export const Mosaic: React.FunctionComponent<Props> = ({images}) => (
  <div>
    {images.map((el, i) => [
      <img
        style={{
          width: '100%',
        }}
        src={`https://source.unsplash.com/1600x${900 + i}`}
        title="Sample title"
        alt={'hello'}
      />,
      <br />,
    ])}
  </div>
);
