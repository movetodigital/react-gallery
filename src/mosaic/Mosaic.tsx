/* eslint-disable @typescript-eslint/no-explicit-any */
import chunk from 'lodash/chunk';
import React, {useState} from 'react';
import styled from 'styled-components';

type FunctionName = (n: object) => any;

interface ImageItemProps {
  src?: string;
  title?: string;
  alt?: string;
}

interface DetailItemProps {
  src?: string;
  title?: string;
  alt?: string;
  id?: string;
  width?: number;
}

interface ComponentProps {
  images: Array<ImageItemProps>;
  gridColumns: number;
  gridGutter: number;
  aspectRatio?: number;
  imageRenderer?: FunctionName;
  enableDetailView?: boolean;
  detailsViewRenderer?: FunctionName;
}

interface StyleItemProps {
  readonly width: number;
  readonly padding: string;
  readonly ratio: number;
}

interface RendererProps {
  props: ImageItemProps;
}
interface DetailViewRendererProps {
  props: DetailItemProps;
}
const Images = styled.div`
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
`;

const Item = styled.div<StyleItemProps>`
  color: white;
  position: relative;
  box-sizing: border-box;
  width: ${props => props.width}%;
  margin: ${props => props.padding};

  > div {
    width: 100%;
    padding-top: ${props => props.ratio}%;
  }
`;

const DefaultImage = styled.img`
  width: 100%;
  height: 100%;
  object-position: 50% 50%;
  object-fit: cover;
`;

const DefaultImageContainer = styled.div`
  position: absolute;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
`;

const DetailView = styled.div`
  position: relative;
`;

const DefaultDetailContainer = styled.div<DetailItemProps>`
  width: ${props => props.width}%;
  margin-left: auto;
  margin-right: auto;
`;
const defaultRender = ({props: {src, title, alt}}: RendererProps) => {
  return (
    <DefaultImageContainer>
      <DefaultImage key={src} src={src} title={title} alt={alt} />
    </DefaultImageContainer>
  );
};

const defaultDetailsViewRenderer = ({
  props: {src, title, alt, width},
}: DetailViewRendererProps) => {
  return (
    <DefaultDetailContainer width={width}>
      <DefaultImage key={src} src={src} title={title} alt={alt} />
    </DefaultDetailContainer>
  );
};

const Mosaic = ({
  images = [],
  gridColumns = 5,
  gridGutter = 1,
  aspectRatio = 1,
  imageRenderer = defaultRender,
  enableDetailView = false,
  detailsViewRenderer = defaultDetailsViewRenderer,
}): ComponentProps => {
  const gutter = gridGutter / 2;
  const width = (100 - (gridColumns - 1) * gutter * 2) / gridColumns;
  const ratio = 100 / aspectRatio;

  const [selectedImageRow, setSelectedImageRow] = useState(null);
  const [selectedImage, setSelectedImage] = useState(null);
  const [selectedImageId, setSelectedImageId] = useState(null);

  const handleSelectImage = ({imageRow, imageId, image}: any) => {
    if (selectedImageRow !== imageRow || imageId !== selectedImageId) {
      setSelectedImageRow(imageRow);
      setSelectedImage(image);
      setSelectedImageId(imageId);
    } else {
      setSelectedImageRow(null);
      setSelectedImage(null);
      setSelectedImageId(null);
    }
  };

  // @ts-ignore
  return chunk(images, gridColumns).map((row: Array<any>, rowIndex: number) => {
    return (
      <div key={`image-row-${rowIndex}`}>
        <Images>
          {row.map((el: any, index: number) => {
            const imageId = `image-${el.src}-${rowIndex}-${index}`;
            return (
              <Item
                key={imageId}
                width={width}
                padding={`${gutter}% ${
                  (index + 1) % gridColumns === 0 ? '0' : gutter
                }% ${gutter}% ${index % gridColumns === 0 ? '0' : gutter}%`}
                ratio={ratio}
                onClick={() =>
                  handleSelectImage({
                    imageRow: rowIndex,
                    imageId,
                    image: el,
                  })
                }
              >
                <div>{imageRenderer({props: el})}</div>
              </Item>
            );
          })}
        </Images>
        <DetailView>
          {enableDetailView &&
            rowIndex === selectedImageRow &&
            detailsViewRenderer({
              props: {
                width,
                // @ts-ignore
                ...selectedImage,
              },
            })}
        </DetailView>
      </div>
    );
  });
};

export {Mosaic};
