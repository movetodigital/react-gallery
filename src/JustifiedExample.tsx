import * as React from 'react';

interface Props {
  margin?: number;
}

import {Justified} from './Justified';

const getRandomInt = (max: number) => {
  return Math.floor(Math.random() * Math.floor(max));
};

const images = Array.from('x'.repeat(40)).map(item => ({
  src: `https://source.unsplash.com/800x${800 + getRandomInt(100)}`,
  title: 'hello',
  alt: 'hello',
}));
export const JustifiedExample: React.FunctionComponent<Props> = ({margin}) => (
  <div style={{width: '1200px', margin}}>
    <Justified
      aspectRatio={16 / 9}
      gridColumns={5}
      gridGutter={1}
      images={images}
    />
  </div>
);
