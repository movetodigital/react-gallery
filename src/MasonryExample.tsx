import * as React from 'react';

interface Props {
  margin?: number;
}

import {Masonry} from './Masonry';

const getRandomInt = (max: number) => {
  return Math.floor(Math.random() * Math.floor(max));
};

const images = Array.from('x'.repeat(40)).map(item => ({
  src: `https://source.unsplash.com/800x${800 + getRandomInt(700)}`,
  title: 'hello',
  alt: 'hello',
}));
export const MasonryExample: React.FunctionComponent<Props> = ({margin}) => (
  <div style={{width: '1200px', margin}}>
    <Masonry gridColumns={5} gridGutter={1} images={images} />
  </div>
);
