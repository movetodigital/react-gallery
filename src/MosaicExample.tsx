import * as React from 'react';

interface Props {
  margin?: number;
}

import {Mosaic} from './Mosaic';

const images = Array.from('x'.repeat(40));
export const MosaicExample: React.FunctionComponent<Props> = ({margin}) => (
  <div style={{maxWidth: '400px', margin}}>
    <Mosaic images={images} />
  </div>
);
