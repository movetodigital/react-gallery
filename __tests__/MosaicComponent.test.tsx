import {shallow} from 'enzyme';
import * as React from 'react';

import {Mosaic} from '../src/Mosaic';

const hydratedProps = {
  images: [
      'x',
      'x',
      'x',
      'x'
  ]
};

describe('<Mosaic />', () => {
  it('should render and match snapshot', () => {
    const tree = shallow(<Mosaic {...hydratedProps} />);

    expect(tree).toMatchSnapshot();
  });
});
