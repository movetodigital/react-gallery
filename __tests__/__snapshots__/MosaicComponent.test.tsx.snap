// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Mosaic /> should render and match snapshot 1`] = `
<div>
  <img
    alt="hello"
    src="https://source.unsplash.com/1600x900"
    style={
      Object {
        "width": "100%",
      }
    }
    title="Sample title"
  />
  <br />
  <img
    alt="hello"
    src="https://source.unsplash.com/1600x901"
    style={
      Object {
        "width": "100%",
      }
    }
    title="Sample title"
  />
  <br />
  <img
    alt="hello"
    src="https://source.unsplash.com/1600x902"
    style={
      Object {
        "width": "100%",
      }
    }
    title="Sample title"
  />
  <br />
  <img
    alt="hello"
    src="https://source.unsplash.com/1600x903"
    style={
      Object {
        "width": "100%",
      }
    }
    title="Sample title"
  />
  <br />
</div>
`;
